api/maven-api-model/src/main/mdo/maven.mdo

```diff
@@ -1945,6 +1965,178 @@
+            The generated build system will compile these directories when the project is being tested.
+            The path given in the {@code directory} field is relative to the project descriptor.
+            The default directory for the default language (Java) is {@code "src/test/java"}.</p>
+
+            <p>If no scope is specified, the default is {@code main}.</p>
+            ]]>
+          </description>
+          <type>String</type>
+          <defaultValue>main</defaultValue>
+        </field>
+        <field>
+          <name>lang</name>
+          <version>4.1.0+</version>
+          <description>
+            <![CDATA[
+            Specifies the language of the source files - typically {@code java} or {@code resources}.
+            Resources is used as a generic term for scripting languages (e.g., JavaScript or Python)
+            or markup languages (e.g. properties file, <abbr>JSON</abbr> or <abbr>XML</abbr>).
+
+            <p>The <b>java</b> language is used for specifying a directory containing the Java sources of the project.
+            The generated build system will compile the sources from this directory using the Java compiler when the
+            project is built. The path given in the {@code directory} field is relative to the project descriptor.
+            The default directory for the main Java sources is {@code "src/main/java"}.</p>
+
+            <p>The <b>resources</b> language is used for specifying a directory containing the class-path
+            or module-path resources such as properties files or scripts associated with a project.
+            This directory is meant to be different from the main source directory,
+            in that its contents will be copied to the output directory in most cases
+            (since scripts are interpreted rather than compiled).
+            The default directory for the main resources is {@code "src/main/resources"}.</p>
+
+            <p>If no language is specified, the default is {@code java}.</p>
+            ]]>
+          </description>
+          <type>String</type>
+          <defaultValue>main</defaultValue>
+        </field>
+        <field>
+          <name>module</name>
+          <version>4.1.0+</version>
+          <description>
+            <![CDATA[
+            Name of the Java module (or other language-specific module) which is built by the sources.
+            This element can be specified in a Maven project containing multiple Java modules.
+            It is generally not needed for non-modular projects, or for modular projects having only one module.
+
+            <p>If a module name is specified for resources or script files,
+            then this value modifies the directory where the files will be copied.
+            For example, if a Java module name is "foo.biz", then the {@code foo/bar.properties}
+            resource file will be copied as {@code foo.biz/foo/bar.properties}.</p>
```

Comment:
Is that a Maven convention, a Java convention, or something imposed by the JVM ?

Reply:
It is a Java compiler convention. In a multi-modules project, `javac` automatically inserts module names at the root of class directories. For example, if a module named `my.foo.module` contains a `foo.bar.biz.Bla` class, then in single-module project (as today), `javac` generates the following file in the Maven `target/classes/` directory:

* `foo/bar/biz/Bla.class`

But in a multi-modules project (would be optional and not the default), `javac` generates the same file at the following location instead:

* `my.foo.module/foo/bar/biz/Bla.class`

Other Java tools such as `java` understand this directory hierarchy. I.e. the `--module-path` can be a directory of modules, with no need to enumerate on the path each module found in the directory. I'm not sure if it is a requirement that the directory names match module names, but this is what `javac` does at least by default.

Above documentation of `module` element tries to match what `javac` does. It could be rephrased as _"During Java compilation, we let `javac` do whatever it wants. For resources, we try to copy them in the same directory as `javac`"_. The current wording tries to be more specific about what happens, but the goal is to match `javac` and the documentation would be updated if we find mismatches.